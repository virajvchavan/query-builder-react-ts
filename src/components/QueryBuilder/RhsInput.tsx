import React from 'react';
import { Row } from '../../queryConfig/queryConfig';
import Select, { ValueType } from 'react-select';
import StyledInput from '../StyledInput';
import MultiValueInputSelector from '../MultiValueInputSelector';
import { OptionType, JsonType } from './QueryRow';

type rulesType = string[] | undefined;
type rhsType = string | number[] | string[] | undefined;

interface Props {
    rhs: rhsType,
    queryConfig: Row,
    onNormalRhsChange: (evt: React.ChangeEvent<HTMLInputElement>) => void,
    onSelectRhsChange: (newElement: ValueType<OptionType>) => void,
    onMultiSelectNumbersRhsChange: (newRhs: Array<number>) => void
}

export default function RhsInput({rhs, queryConfig, onMultiSelectNumbersRhsChange, onNormalRhsChange, onSelectRhsChange}: Props) {
    let rules = queryConfig.rhs.config?.split(";");
    switch (queryConfig.rhs.type) {
        case "text":
            let stringValue = rhs as string;
            return <StyledInput type="text" onChange={onNormalRhsChange} defaultValue={stringValue} />;
        case "number":
            let value = rhs as string;
            return <NumberInput rules={rules} rhs={value} onNormalRhsChange={onNormalRhsChange} />
        case "multi-select-list":
            let stringArrValue = rhs as Array<string>;
            return <MultiSelectList rhs={stringArrValue} rules={rules} onSelectRhsChange={onSelectRhsChange} />
        case "multi-select-numbers":
            let numbersArrayValue = rhs as Array<number>;
            return <MultiValueInputSelector onChange={onMultiSelectNumbersRhsChange} defaultValues={numbersArrayValue} />;
        default:
            return null;
    }
}

interface MultiSelectListProps {
    rules: rulesType,
    rhs: Array<string>,
    onSelectRhsChange: (newElement: ValueType<OptionType>) => void
}

function MultiSelectList({rules, rhs, onSelectRhsChange}: MultiSelectListProps) {
    let values: Array<OptionType> = [];
    let optionsList: Array<OptionType> = [];
    rules?.forEach(rule => {
        let [ruleName, ruleValue] = rule.split(":");
        if (ruleName === "file") {
            let optionsJson: JsonType = require("../../queryConfig/" + ruleValue);
            if (Array.isArray(rhs) && typeof rhs[0] === "string") {
                rhs?.forEach((item: string | number) => {
                    values.push({ label: optionsJson[item], value: item.toString() });
                });
            }
            optionsList = Object.keys(optionsJson).map(key => { return { value: key, label: optionsJson[key] }; });
        }
    });

    return <Select options={optionsList} onChange={onSelectRhsChange} isMulti={true} defaultValue={values} />;
}

interface NumberInputProps {
    rules: rulesType,
    onNormalRhsChange: (evt: React.ChangeEvent<HTMLInputElement>) => void,
    rhs: string
}

function NumberInput({rules, rhs, onNormalRhsChange}: NumberInputProps) {
    let options: JsonType = {};
    rules?.forEach(rule => {
        let [ruleName, ruleValue] = rule.split(":");
        if (ruleName === "precision") {
            options.step = ruleValue;
        }
    });
    return <StyledInput type="number" onChange={onNormalRhsChange} options={options} defaultValue={rhs} />;
}
