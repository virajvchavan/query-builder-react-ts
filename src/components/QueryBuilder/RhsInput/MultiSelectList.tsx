import React from 'react';
import Select, { ValueType } from 'react-select';
import { OptionType, JsonType } from '../QueryRow';
import { rulesType } from './RhsInput';

interface MultiSelectListProps {
    rules: rulesType;
    rhs: Array<string>;
    onSelectRhsChange: (newElement: ValueType<OptionType>) => void;
}

export default function MultiSelectList({ rules, rhs, onSelectRhsChange }: MultiSelectListProps) {
    let values: Array<OptionType> = [];
    let optionsList: Array<OptionType> = [];
    rules?.forEach(rule => {
        let [ruleName, ruleValue] = rule.split(":");
        if (ruleName === "file") {
            let optionsJson: JsonType = require("../../../queryConfig/" + ruleValue);
            if (Array.isArray(rhs) && typeof rhs[0] === "string") {
                rhs?.forEach((item: string | number) => {
                    values.push({ label: optionsJson[item], value: item.toString() });
                });
            }
            optionsList = Object.keys(optionsJson).map(key => { return { value: key, label: optionsJson[key] }; });
        }
    });

    return <Select options={optionsList} onChange={onSelectRhsChange} isMulti={true} defaultValue={values} />;
}
