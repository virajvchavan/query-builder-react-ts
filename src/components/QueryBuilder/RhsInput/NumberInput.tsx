import React from 'react';
import StyledInput from '../../StyledInput/StyledInput';
import { JsonType } from '../QueryRow';
import { rulesType } from './RhsInput';

interface NumberInputProps {
    rules: rulesType;
    onNormalRhsChange: (evt: React.ChangeEvent<HTMLInputElement>) => void;
    rhs: string;
}
export function NumberInput({ rules, rhs, onNormalRhsChange }: NumberInputProps) {
    let options: JsonType = {};
    rules?.forEach(rule => {
        let [ruleName, ruleValue] = rule.split(":");
        if (ruleName === "precision") {
            options.step = ruleValue;
        }
    });
    return <StyledInput type="number" onChange={onNormalRhsChange} options={options} defaultValue={rhs} />;
}
