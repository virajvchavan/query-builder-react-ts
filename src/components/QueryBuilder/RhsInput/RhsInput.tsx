import React from 'react';
import { Row } from '../../../queryConfig/queryConfig';
import { ValueType } from 'react-select';
import StyledInput from '../../StyledInput/StyledInput';
import MultiValueInputSelector from '../../MultiValueInputSelector/MultiValueInputSelector';
import { OptionType } from '../QueryRow';
import { MultiSelectList } from './MultiSelectList';
import { NumberInput } from './NumberInput';

export type rulesType = string[] | undefined;
type rhsType = string | number[] | string[] | undefined;

interface Props {
    rhs: rhsType,
    queryConfig: Row,
    onNormalRhsChange: (evt: React.ChangeEvent<HTMLInputElement>) => void,
    onSelectRhsChange: (newElement: ValueType<OptionType>) => void,
    onMultiSelectNumbersRhsChange: (newRhs: Array<number>) => void
}

function RhsInput({rhs, queryConfig, onMultiSelectNumbersRhsChange, onNormalRhsChange, onSelectRhsChange}: Props) {
    let rules = queryConfig.rhs.config?.split(";");
    switch (queryConfig.rhs.type) {
        case "text":
            let stringValue = rhs as string;
            return <StyledInput type="text" onChange={onNormalRhsChange} defaultValue={stringValue} />;
        case "number":
            let value = rhs as string;
            return <NumberInput rules={rules} rhs={value} onNormalRhsChange={onNormalRhsChange} />
        case "multi-select-list":
            let stringArrValue = rhs as Array<string>;
            return <MultiSelectList rhs={stringArrValue} rules={rules} onSelectRhsChange={onSelectRhsChange} />
        case "multi-select-numbers-in-a-range":
            let numbersArrayValue = rhs as Array<number>;
            return <MultiValueInputSelector type="number" onChange={onMultiSelectNumbersRhsChange} defaultValues={numbersArrayValue} />;
        default:
            return null;
    }
}

export default RhsInput;
