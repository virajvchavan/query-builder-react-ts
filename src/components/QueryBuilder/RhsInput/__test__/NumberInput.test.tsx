import { render } from '@testing-library/react';
import React from "react";
import userEvent from '@testing-library/user-event'
import { NumberInput } from "../NumberInput";
import { rulesType } from '../RhsInput';

test("it should render an input with type=number", () => {
    const { getByTestId } = render(<NumberInput rules={[]} value="" options={{placeholder: "Type a number"}} onNormalRhsChange={jest.fn()} />);
    expect(getByTestId("rhsInput")).toHaveAttribute("type", "number");
});

test("it should apply options specified as attributed", () => {
    const { getByTestId } = render(<NumberInput rules={[]} value="" options={{placeholder: "Type a number"}} onNormalRhsChange={jest.fn()} />);
    expect(getByTestId("rhsInput")).toHaveAttribute("placeholder", "Type a number");
});

test("it should apply correct precision values for number input", () => {
    let rules: rulesType = ["precision:0.01"];
    const { getByPlaceholderText } = render(<NumberInput rules={rules} value="" options={{placeholder: "Type a number"}} onNormalRhsChange={jest.fn()} />);
    let inputElement = getByPlaceholderText("Type a number");
    expect(inputElement).toHaveAttribute("step", "0.01");
});

test("it should call onChange function when user types", () => {
    let onChange = jest.fn();
    const { getByTestId } = render(<NumberInput rules={[]} value="" options={{placeholder: "Type a number"}} onNormalRhsChange={onChange} />);
    let inputElement = getByTestId("rhsInput");
    userEvent.type(inputElement, "5");
    expect(onChange).toHaveBeenCalledTimes(1);
});
