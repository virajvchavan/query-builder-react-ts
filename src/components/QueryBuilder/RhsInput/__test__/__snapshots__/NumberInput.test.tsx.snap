// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should match snapshot for the initial render 1`] = `
<DocumentFragment>
  <input
    class="styled-input"
    data-testid="rhsInput"
    placeholder="Type a number"
    type="number"
    value=""
  />
</DocumentFragment>
`;
